### PR TITLE
Phosphorizer: "a la carte" light tube/bulb recoloring machine

### PR DESCRIPTION
**Pull request by Kubius**
[FEAT]
## About the PR 

phosphorizerC

Adds (but does not yet introduce, see end note) the **Phosphorizer**, a large machine capable of applying custom-blended phosphors to light tubes, changing the color of the light they output. The machine has access to nearly all of the RGB color spectrum (small lower limit), can be dialed per base color, accepts lights from tube/bulb boxes, and includes a preview swatch.

## Why's this needed? 

Provides engineering personnel (or other people intruding on the department) with much more granularity in their choice of lighting for construction projects.

Addition to the Engineering department of maps will happen in a subsequent PR if this is accepted, at which point a changelog entry would also be included.

### CODE REVIEW

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        Please insert only standard light tubes and bulbs.
+      </NoticeBox>
+      <Box>
+        <strong># Loaded Lights:</strong>
+        {' '}{ tubes }
+      </Box>
+      <Box>
+        Color Tuning:
+        {' '}
+        <NumberInput
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Swapped to colored knobs in 75a888f

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        <ColorBox
+          color={`rgba(${hostR}, ${hostG}, ${hostB}, 1)`} />
+      </Box>
+      <Divider />
+      <Button
+        icon="power-off"
+        content={online ? "Stop Phosphorizing" : "Start Phosphorizing"}
+        tooltip={online ? "Abort current processing." : "Begin applying chosen color to contents."}
+        tooltipPosition="right"
+        onClick={() => act('process', {})} />
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Seems reasonable and removal of second argument in 75a888f didn't yell at me, so thumbs up.

File: code/obj/machinery/phosphorizer.dm [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+			. = TRUE
+		if("process")
+			if(status | BROKEN && powered() && src.contents.len)
+				if(!src.phosphorizing)
+					src.start_phos()
+					boutput(usr, "You activate [src].")
+				else
+					src.stop_phos()
+					boutput(usr, "You deactivate [src].")
+				ui_interact(usr, ui)
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm]: Replaced in 75a888f

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
+import { useBackend } from '../backend';
+import { Box, Button, NoticeBox, Divider, ColorBox, NumberInput } from '../components';
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Alphabetized in 75a888f but differently to shown form (removed NumberInput and added Knob as a consequence of other changes).

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          )}
+        </Box>
+      </Window.Content>
+    </Window>
+  );
+};
+
+const OperateWindow = (props, context) => {
+  const { act, data } = useBackend(context);
+  const { tubes, online, hostR, hostG, hostB } = data;
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Sensible enough. Changed in 75a888f

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+          value={hostB}
+          format={value => "B:" + value}
+          onDrag={(e, value) => act('tune_hue', {
+            hue: "B",
+            output: value,
+          })}
+        />
+        {' '}
+        <ColorBox
+          color={`rgba(${hostR}, ${hostG}, ${hostB}, 1)`} />
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Inlined in 75a888f

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      </Box>
+      <Divider />
+      <Button
+        icon="power-off"
+        content={online ? "Stop Phosphorizing" : "Start Phosphorizing"}
+        tooltip={online ? "Abort current processing." : "Begin applying chosen color to contents."}
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Sure, why not. Added in 75a888f

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        tooltipPosition="bottom"
+        onClick={() => act('eject', {})} />
+    </Box>
+  );
+};
+
+const BusyWindow = () => {
+  return (
+    <NoticeBox warning>
+      The Machine is busy, please wait!
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Adjusted in 75a888f to a different and more "flavorful" busy notice.

File: code/obj/machinery/phosphorizer.dm [hunk 9]
```diff
@@ -0,0 +1,163 @@
+/obj/machinery/phosphorizer
+	name = "Phosphorizer"
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm]: Had a vague idea that it's a proper name, but given the lack of a company with which it's associated, it probably makes more sense as generic. Done in 75a888f.

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+export const Phosphorizer = (props, context) => {
+  const { data } = useBackend(context);
+  const { busy } = data;
+  return (
+    <Window
+      title="Phosphorizer"
+      width={360}
+      height={190}>
+      <Window.Content>
+        <Box>
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Deboxed in 75a888f.

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        content={online ? "Stop Phosphorizing" : "Start Phosphorizing"}
+        tooltip={online ? "Abort current processing." : "Begin applying chosen color to contents."}
+        tooltipPosition="right"
+        onClick={() => act('process', {})} />
+      <Button
+        icon="eject"
+        content="Eject"
+        tooltip="Remove contents without colorizing."
+        tooltipPosition="bottom"
+        onClick={() => act('eject', {})} />
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Done in 75a888f.

**Comment by Kubius** [comment]
Updated UI, as of 75a888f

image

**Comment by ZeWaka** [comment]
> Updated UI, as of [75a888f](https://github.com/goonstation/goonstation/commit/75a888f8542202e45b0b8924d1eea1c5be4de576)
> 
> image

What's with the weird pound sign?

**Comment by Kubius** [comment]
It's an indicator of the following number being a count of things (since it's a number sign) - essentially an abbreviated form of "Number of".

File: code/obj/machinery/phosphorizer.dm [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+		return
+	switch(action)
+		if("tune_hue")
+			switch(params["hue"])
+				if("R") src.ctrl_R = clamp(round(params["output"]), 20, 255)
+				if("G") src.ctrl_G = clamp(round(params["output"]), 20, 255)
+				if("B") src.ctrl_B = clamp(round(params["output"]), 20, 255)
+			. = TRUE
+		if("toggle-process")
+			if(status | BROKEN && powered() && src.contents.len)
```
Comment (mordent-goonstation) [inline code/obj/machinery/phosphorizer.dm:146]:
```suggestion
			if (length(src.contents))
```
`ui_act` is bombs out earlier than this [here](https://github.com/goonstation/goonstation/blob/master/code/modules/tgui/external.dm#L78) if `ui.status != UI_INTERACTIVE`.

Machinery status is (by default) determined [here](https://github.com/goonstation/goonstation/blob/master/code/obj/machinery.dm#L186), which uses a "not broken" check that `/obj/machinery` has [here](https://github.com/goonstation/goonstation/blob/master/code/modules/tgui/states/not_broken.dm#L15). That part checks for whether it's broken, and whether (if it needs power) if it has power.

Basiclally: the checks here are superfluous, so I suggested them removed.

File: code/obj/machinery/phosphorizer.dm [hunk 13]
```diff
@@ -0,0 +1,163 @@
+/obj/machinery/phosphorizer
+	name = "phosphorizer"
+	desc = "A device capable of baking a phosphor onto light tubes or bulbs, changing the color of light they emit."
+	icon = 'icons/obj/machines/phosphorizer.dmi'
+	icon_state = "baseunit"
+	anchored = 1
+	density = 1
+	power_usage = 150
```
Comment (mordent-goonstation) [inline code/obj/machinery/phosphorizer.dm:8]:
```suggestion
	power_usage = 150
	status = REQ_PHYSICAL_ACCESS
```
Given that you've gone for a "physical knobs" style interface, I think making it require physical access seems right.

File: code/obj/machinery/phosphorizer.dm [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+			if(status | BROKEN && powered() && src.contents.len)
+				if(!src.phosphorizing)
+					src.start_phos()
+					boutput(usr, "You activate [src].")
+				else
+					src.stop_phos()
+					boutput(usr, "You deactivate [src].")
+				. = TRUE
+		if("eject")
+			if(src.contents.len)
```
Comment (mordent-goonstation) [inline code/obj/machinery/phosphorizer.dm:155]:
```suggestion
			if(length(src.contents))
```
Prefer `length(foo)` over `foo.len`.

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+          color="blue"
+          format={value => "B:" + value}
+          onDrag={(e, value) => act('tune_hue', {
+            hue: "B",
+            output: value,
+          })}
+        />
+        {' '}
+        <ColorBox color={`rgba(${hostR}, ${hostG}, ${hostB}, 1)`} />
+      </Box>
```
Comment (mordent-goonstation) [inline tgui/packages/tgui/interfaces/Phosphorizer.js:78]:
```suggestion
      <Stack vertical>
        <Stack.Item>
          <strong># Loaded Lights:</strong>
          {' '}{ lights }
        </Stack.Item>
        <Stack.Item>
          Color Tuning:
          {' '}
          <Knob
            inline
            minValue={20}
            maxValue={255}
            value={hostR}
            color="red"
            format={value => "R:" + value}
            onDrag={(e, value) => act('tune_hue', {
              hue: "R",
              output: value,
            })}
          />
          <Knob
            inline
            minValue={20}
            maxValue={255}
            value={hostG}
            color="green"
            format={value => "G:" + value}
            onDrag={(e, value) => act('tune_hue', {
              hue: "G",
              output: value,
            })}
          />
          <Knob
            inline
            minValue={20}
            maxValue={255}
            value={hostB}
            color="blue"
            format={value => "B:" + value}
            onDrag={(e, value) => act('tune_hue', {
              hue: "B",
              output: value,
            })}
          />
          {' '}
          <ColorBox color={`rgba(${hostR}, ${hostG}, ${hostB}, 1)`} />
        </Stack.Item>
      </Stack>
```
Prefer using a `Stack` here for better spacing (would need to be imported as well).

**Review by mordent-goonstation — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/Phosphorizer.js (hunk 15 above)
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js]: Stackified in f7358dc.

File: code/obj/machinery/phosphorizer.dm (hunk 12 above)
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm]: Removal done in f7358dc.

File: code/obj/machinery/phosphorizer.dm (hunk 14 above)
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm]: Yep, sure. No instances of .len in the file now. f7358dc, as with the rest of this batch.

File: code/obj/machinery/phosphorizer.dm (hunk 13 above)
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm:8]: Seems sensible. f7358dc

File: code/obj/machinery/phosphorizer.dm [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+			sleep(phos_delay)
+
+		if(failbreak)
+			failbreak = false
+			src.visible_message("<b>[src]</b> abruptly stops operating.")
+			playsound(src.loc, sound_grump, 20, 1)
+		else
+			src.visible_message("<b>[src]</b> finishes working and shuts down.")
+			playsound(src.loc, sound_happi, 20, 1)
+		if(src.phosphorizing) stop_phos()
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm]: Retooled in f7358dc to be machine loop based, seems to work ingame (tested regular operation, ejection during processing, power out during processing).

**Comment by Kubius** [comment]
Retool 2 added machine loop based processing, among other things.

Adjusted art a bit more; powered-on panel (top right) is now darker when not being illuminated by active power, outlines on left section have been softened using an existing dark color, top segments have been darkened a smidge to emphasize the front. Not quite sure about it, but it seems better than the last iteration.

phosphorizerD_96

**Comment by ZeWaka** [comment]
> It's an indicator of the following number being a count of things (since it's a number sign) - essentially an abbreviated form of "Number of".

why not use №

**Comment by Kubius** [comment]
Haven't seen that used ingame anywhere, or in real life on anything made after the turn of the century.

**Comment by mordent-goonstation** [comment]
> To an untrained eye, it looks like "NO", and "NO Loaded Lights" is more easily misinterpreted as a real thing than "hashtag loaded lights" or "pound loaded lights". If anything, I'd switch to just explicitly saying "Number of" if there's a concern over comprehension, but I'd like someone else to sound off on the topic first.

I would honestly just say "Loaded Lights: 0" or "Loaded Lights: 17" or whatever.

**Comment by Kubius** [comment]
Broad support for that approach after checking in with users, so I'll go with it.

**Comment by Flaborized** [comment]
Updated version of sprite isn't that much better IMO, still looks quite a bit like an older asset

**Comment by ZeWaka** [comment]
yeah there's still the wack gradient feel

**Comment by Kubius** [comment]
One more revision of the same fundamental art with some additional alterations (right section adjustment, smoothed out the tops so they don't look so retro). If this is inadequate, which I fear it might be, I'll probably have to entirely redo the art from scratch; trying to convert the same rough framework to the 5-color standard palette lost too much detail.

phosphorizerF

**Comment by Kubius** [comment]
New-palette rework developed collaboratively. If this is suitable, I'll re-kerfuffle the art accordingly.

phosphorework3

File: code/obj/machinery/phosphorizer.dm [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+		else if(load_bulb(W,user))
+			boutput(user, "You load [W] into [src].")
+			playsound(src, sound_load, 40, 1)
+			attack_hand(user)
+		else
+			. = ..()
+
+	proc/load_bulb(obj/item/W as obj, mob/user as mob)
+		. = FALSE
+		if (istype_exact(W,/obj/item/light/tube) || istype_exact(W,/obj/item/light/bulb)) //exact type so you're not coloring a colored bulb
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm:49]: Colors applied by way of phosphorizer are also able to be stripped and reapplied by the phosphorizer, whereas regular colored tubes have it "baked in".

File: code/obj/machinery/phosphorizer.dm [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+		src.phosphorizing = true
+		var/image/O_panel = SafeGetOverlayImage("operatebar", 'icons/obj/machines/phosphorizer.dmi', "activelight")
+		UpdateOverlays(O_panel, "operatebar", 0, 1)
+
+/obj/machinery/phosphorizer/process()
+	if(src.phosphorizing)
+		if(status & NOPOWER)
+			stop_phos()
+		if(length(src.contents))
+			src.visible_message(length(src.contents))
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm:93]: Debug thing. Will delete in next change set.

File: code/obj/machinery/phosphorizer.dm [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+	I_panel.alpha = 220
+	if (status & BROKEN)
+		UpdateOverlays(null, "statuspanel", 0, 1)
+		if(src.phosphorizing) src.stop_phos()
+	else
+		if ( powered() )
+			UpdateOverlays(I_panel, "statuspanel", 0, 1)
+			status &= ~NOPOWER
+		else
+			SPAWN_DBG(rand(0, 5))
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm:117]: Staggered power-off is actually implemented like this in several locations, but I'm open to putting in some sort of failsafe. Would rechecking powered() at the time of the display power-off be appropriate?

File: code/obj/machinery/phosphorizer.dm [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+					src.start_phos()
+					boutput(usr, "You activate [src].")
+				else
+					src.stop_phos()
+					boutput(usr, "You deactivate [src].")
+				. = TRUE
+		if("eject")
+			if(length(src.contents))
+				if(src.phosphorizing)
+					src.visible_message("<b>[src]</b> ejects its unprocessed lights and shuts down.")
```
[the remark this answers is not in the document]
Comment (Kubius) [reply code/obj/machinery/phosphorizer.dm:159]: After a brief search, 1,186 instances of <b>[src]</b>, broadly used to emphasize an entity doing a thing. Not sure it's strictly necessary, but it seemed suitable.

File: tgui/packages/tgui/interfaces/Phosphorizer.js [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        <Stack.Item>
+          <strong>Loaded Lights:</strong>
+          {' '}{ lights }
+        </Stack.Item>
+        <Stack.Item>
+          Color Tuning:
+          {' '}
+          <Knob
+            inline
+            minValue={20}
```
[the remark this answers is not in the document]
Comment (Kubius) [reply tgui/packages/tgui/interfaces/Phosphorizer.js:43]: P sensible. Will do this in next change set.

**Comment by Kubius** [comment]
Haven't had the time to continue polishing this, closing pending further upgrades.

**Pull request closed without merging**